### PR TITLE
tracer: request Headers as tags for web integrations

### PR DESCRIPTION
**Pull request by mtoffl01**
### What does this PR do?
Makes it possible to set request headers as tags at the integration level and at the global tracer level.

### Motivation
Feature parity between tracers

### Reviewer's Checklist

- [ ] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [ ] Changed code has unit tests for its functionality.
- [ ] If this interacts with the agent in a new way, a system test has been added.

### CODE REVIEW

File: contrib/go-chi/chi.v5/chi.go
Comment (mtoffl01) [inline contrib/go-chi/chi.v5/chi.go:15]: **QTNA**: Is this a fair import name?

File: ddtrace/tracer/util_test.go
Comment (mtoffl01) [inline ddtrace/tracer/util_test.go:135]: See: https://datadoghq.atlassian.net/wiki/spaces/APMINT/pages/2302444638/DD+TRACE+HEADER+TAGS#Tag-name
Unclear (to me) what the specifications are around casing.

File: ddtrace/tracer/util.go
Comment (mtoffl01) [inline ddtrace/tracer/util.go:135]: btw, I made the executive decision here that we only expect `header` or `header:tag` as input. That is, if a user passes in something like:
`header:tag:somethingelse`
`somethingelse` will be ignored entirely. Only `header` and `tag` are considered.

File: contrib/emicklei/go-restful/restful_test.go
Comment (katiehockman) [inline contrib/emicklei/go-restful/restful_test.go:50]:
```suggestion
```

File: contrib/go-chi/chi.v5/chi.go
Comment (katiehockman) [reply contrib/go-chi/chi.v5/chi.go:15]: Since other imports of `"gopkg.in/DataDog/dd-trace-go.v1/contrib/net/http"` alias to `httptrace` perhaps we should keep that consistent, and just alias `"gopkg.in/DataDog/dd-trace-go.v1/contrib/internal/httptrace"` to `httptraceinternal` instead

File: contrib/net/http/trace.go
Comment (katiehockman) [inline contrib/net/http/trace.go:109]: Why have this function here, instead of in `gopkg.in/DataDog/dd-trace-go.v1/contrib/internal/httptrace`?
I ask because the package description for internal/httptrace is
> // Package httptrace provides functionalities to trace HTTP requests that are commonly required and used across contrib/** integrations.

This would also eliminate the need to import both internal/httptrace and contrib/net/http within all of those contrib packages, which is making the import aliasing a bit messy.

Comment (katiehockman) [reply contrib/net/http/trace.go:109]: Same question with `ConvertHeaderToTag`. Notice that none of the other functions in `ddtrace/tracer/util.go` are exported. For `ConvertHeaderToTag`, it definitely seems like having the code in an internal shared utility package makes more sense. Otherwise, we're exposing that function as part of the public API, which won't have any use for consumers of our package.

File: ddtrace/tracer/util.go
Comment (katiehockman) [reply ddtrace/tracer/util.go:135]: If someone gives `header:tag:somethingelse`, I would actually say that it should return `"header", "tag:something"`. Though considering all of the other tracers have implemented this, we should just verify how they do it. Ideally we already have system tests that check this which you can verify with, and if we don't, we should add one to see what the other tracers do.

I would do something like this instead:

```
headerAsTag = strings.ToLower(strings.TrimSpace(headerAsTag))
s := strings.SplitN(headerAsTag, ":", 2)
if len(s) == 1 {
  // not a custom tag
  return s[0], strings.Join(ext.HTTPRequestHeaders, s[0], "."), nil
}
return s[0], s[1]
```

File: contrib/net/http/trace.go
Comment (katiehockman) [inline contrib/net/http/trace.go:107]:
```suggestion
		for h, v := range req.Header {
			h = strings.ToLower(h)
			if tag, ok := headersAsTags[h]; ok && !strings.HasPrefix(h, "x-datadog-") {
				cfg.Tags[tag] = strings.Join(v, ",")
			}
		}
```
Would this work instead? No need to call out `req.Header.Values(k)` if you already have them as the second argument of the range (which is currently just being thrown away)

**Review by katiehockman — COMMENTED** [review]
We should double check that we don't need to do anything with the Header name normalization which is described in the design doc.

File: contrib/go-chi/chi.v5/chi.go
Comment (mtoffl01) [reply contrib/go-chi/chi.v5/chi.go:15]: That makes sense to me -- should I do that in this PR, or should we separate it?

File: ddtrace/tracer/util.go
Comment (mtoffl01) [reply ddtrace/tracer/util.go:135]: AFAICT, the system test actually does the same thing I'm doing — it looks at the before and after of the first colon, and nothing else: https://github.com/DataDog/system-tests/blob/a965a8f6f5bf143da916152faf613809978774fc/tests/test_library_conf.py#L64

File: contrib/net/http/trace.go
Comment (mtoffl01) [reply contrib/net/http/trace.go:109]: Note for myself as I look into this:

- originally chose to put HeaderTagsFromRequest in here b/c I though there were some import issues importing httptrace into the net/http pkg, but let me revisit this
- ConvertHeaderToTag is used in the global tracer package b/c the global config also needs access to it, hence why I put it in tracer/util rather than a contrib pkg, but will revisit this

Comment (mtoffl01) [reply contrib/net/http/trace.go]: Additional notes:

- no circular import issues, actually. However, mux integration currently imports the net/http pkg with import name "httptrace", so we will have to change some import names across the board anyway (i.e, either change net/http import names away from httptrace so that the httptrace pkg can use it, or rename httptrace pkg/ rename httptrace pkg imports to something else e.g, httptraceinternal
-

**Comment by Julio-Guerra** [comment]
I don't like the overall httptrace -> ddhttp renaming of this PR, while we always used to name this package (and others) this `*trace` way. This name is also used in our documentation for example.
Is this intended and necessary (I don't see it related to the request headers tags)? If not, I suggest reverting this change.

**Comment by mtoffl01** [comment]
> I don't like the overall httptrace -> ddhttp renaming of this PR, while we always used to name this package (and others) this `*trace` way. This name is also used in our documentation for example. Is this intended and necessary (I don't see it related to the request headers tags)? If not, I suggest reverting this change.

@Julio-Guerra , this is _related_ to trace header tags because some of the functionality for this feature lives in the httptrace pkg. Here's why I re-aliased the contrib/net/http pkg from httptrace to ddhttp:

There are cases, like in the `mux` pkg, where I am importing all of the following:
- "net/http"
- "gopkg.in/DataDog/dd-trace-go.v1/contrib/internal/httptrace"
- "gopkg.in/DataDog/dd-trace-go.v1/contrib/net/http"

Historically, the contrib/net/http has been aliased to `httptrace` to avoid conflict with the standard net/http library. But given that we are also importing a package whose name is actually httptrace -- that is, contrib/internal/httptrace -- I figured the "rightful owner" of the `httptrace` alias was the httptrace package itself. Thus I re-aliased the contrib/net/http package.

The alternative is to keep contrib/net/http's standard alias as httptrace but then we must either (1) re-alias the httptrace package (this seems very confusing when contrib/net/http and contrib/internal/httptrace are both imported), or (2) rename the httptrace package, maybe to something like httptraceinternal ? But 2 might cause an even bigger rupture than we have now. LMK what you think!

**Comment by Julio-Guerra** [comment]
> > I don't like the overall httptrace -> ddhttp renaming of this PR, while we always used to name this package (and others) this `*trace` way. This name is also used in our documentation for example. Is this intended and necessary (I don't see it related to the request headers tags)? If not, I suggest reverting this change.
> 
> @Julio-Guerra , this is _related_ to trace header tags because some of the functionality for this feature lives in the `httptrace` pkg. Here's why I re-aliased the contrib/net/http pkg from `httptrace` to `ddhttp`:
> 
> There are cases, like in the `mux` pkg, where I am importing all of the following:
> 
> * "net/http"
> * "gopkg.in/DataDog/dd-trace-go.v1/contrib/internal/httptrace"
> * "gopkg.in/DataDog/dd-trace-go.v1/contrib/net/http"
> 
> Historically, the contrib/net/http has been aliased to `httptrace` to avoid conflict with the standard net/http library. But given that we are also importing a package whose name is actually `httptrace` — that is, contrib/internal/httptrace — I figured the "rightful owner" of the `httptrace` alias was the `httptrace` package itself. Thus I re-aliased the contrib/net/http package.
> 
> The alternative is to keep contrib/net/http's standard alias as `httptrace` but then we must either (1) re-alias the `httptrace` package (this seems very confusing when contrib/net/http and contrib/internal/httptrace are both imported), or (2) rename the `httptrace` package, maybe to something like `httptraceinternal` ? But 2 might cause an even bigger rupture than we have now. LMK what you think!

I see your point indeed. But I'd still favor and honor this [unofficial] public naming convention we use in all our examples. This internal httptrace package name conflict that you state is only local to our Go module and doesn't impact external users.

I let the @DataDog/tracing-go team decide in the end and I'll approve the @DataDog/appsec-go change (which is only about this package alias name change) if they do accept it and want to move this way from now on.

**Comment by mtoffl01** [comment]
This PR enables automatic request header tagging on spans for the following contrib pkgs:

- emicklei/go-restful
- gin-gonic/gin
- go-chi/chi
- go-chi/chi.v5
- gorilla/mux
- net/http
- urfave/negroni

I decided on these integrations because they all call on `httptrace.StartRequestSpan`. However, there are other integrations which appear to make http.request spans, such as julienschmidt/httprouter which are not [yet] included in this list. I would like input on whether the list above is complete [for now], or if there are other integrations that we should add support for in this PR.

This PR may also be expanded to include response header tags as well (this is pending input/review from other members of the team, see: https://datadoghq.atlassian.net/browse/AIT-6331)

File: internal/normalizer/normalizer.go
Comment (katiehockman) [inline internal/normalizer/normalizer.go:27]: Suggested refactor for readability: https://go.dev/play/p/MczD7Lshfz9
  - If you know it's going to have to be lowercase, then just make the entire string lowercase at the beginning of the function. i.e. first line is `headerAsTag = strings.ToLower(headerAsTag)`. Then you don't need to call it several times in this fn, and you can drop the capital letters from the regexp
  - I found the if statement on the index a bit difficult to parse, and some of the lines a bit too long.

```
func normalize(headerAsTag string) (header string, tag string) {
	header = strings.ToLower(strings.TrimSpace(headerAsTag))
	if last := strings.LastIndex(header, ":"); last >= 0 {
		// Split the tag from the colon if one was found.
		header, tag = header[:last], header[last+1:]
		header, tag = strings.TrimSpace(header), strings.TrimSpace(tag)
		tag = headerTagRegexp.ReplaceAllString(tag, "_") // replace invalid characters
	} else {
		tag = fmt.Sprintf("http.%s", headerTagRegexp.ReplaceAllString(header, "_"))
	}
	return header, tag
}
```

Comment (mtoffl01) [reply internal/normalizer/normalizer.go:27]: @katiehockman , I specifically did not perform the lowercasing before the parsing because of this statement in the spec: 
> NB: if the user specifies a mapping part cf [above configuration 2)](https://datadoghq.atlassian.net/wiki/spaces/APMINT/pages/2302444638/DD+TRACE+HEADER+TAGS#mapping-the-header), no normalization should happen. The user is responsible for the mapping part, if some nesting is created unintentionally then configuration shoudl be fixed.

I took this to mean that if a user includes a mapping part, that this should remain untouched (except for removal of leading/trailing whitespaces after the colon & before the end of the string). LMK if you still think that lowercasing should join whitespacing though (Or alternatively that "0 normalization" = truly nothing, i.e, _no_ trimming of whitespaces?)

Comment (mtoffl01) [reply internal/normalizer/normalizer.go:27]: TODO: Run parametric tests to check how other libraries handle this.

Comment (mtoffl01) [reply internal/normalizer/normalizer.go:27]: @katiehockman , in my original logic, I was handling the following cases as if they include no tag mapping / as if the colon was merely a mistake:
`:header`
`header:`
That is, I was treating both cases as if the input was simply "header", no tag mapping specified.

But the logic you provided in this thread treats the above as input with valid tag mapping.
- For input ":header", header "" gets mapped to tag "header"
- For input "header:", header "header" gets mapped to tag ""

I can also try to use the parametric tests to check how the other libraries handle this input, because it is up for debate at this point.

**Comment by mtoffl01** [comment]
Want to make sure reviewers see this comment: https://github.com/DataDog/dd-trace-go/pull/1764#issuecomment-1466719571

**Review by ajgajg1134 — CHANGES_REQUESTED** [review]
Overall this is looking good! Just a few small nits :)

File: internal/normalizer/normalizer.go
Comment (mtoffl01) [reply internal/normalizer/normalizer.go:15]: @ajgajg1134 , note that the regex actually allows spaces and dashes, but any non-alphanumerics, non-spaces, non-dashes will get replaced with underscores.

My decision not to normalize dashes comes from this example in the RFC:

And my decision not to normalize spaces comes from this conclusion in my [system tests analysis doc](https://docs.google.com/document/d/1zP0NugQCSq-J23dERbdVTcQws8rUR8Pjpp0AV69kVKc/edit#heading=h.lddbbn5byimh)

> Conclusions drawn from HTTP specs:
> leading whitespaces not allowed in header name field
> trailing whitespaces not allowed in header name field
> between-char whitespaces allowed in header name field
and
> [The Datadog RFC] does not specify normalization (or lack thereof) for between-char whitespaces. With this in mind, I’ve created tests to ensure that leading and trailing whitespaces are essentially trimmed from the DD_TRACE_HEADER_TAGS input, but between-char spaces are left in-tact. 

So instead I'm updating the description of `headerTagRegexp` to be:
// headerTagRegexp is used to replace all invalid characters in the config. Only alphanumerics, whitespaces and dashes allowed.

Comment (ajgajg1134) [reply internal/normalizer/normalizer.go:15]: oh whoops :P, now I'm extra glad we're updating the comment since I even misunderstood it in my original edit. 👍

**Review by ajgajg1134 — APPROVED** [review]
Looks good to me!

File: internal/globalconfig/globalconfig.go
Comment (nsrip-dd) [inline internal/globalconfig/globalconfig.go:28]: I notice that access to this isn't protected by `mu`. Do we ever expect SetHeaderTag/GetHeaderTag/etc. to be called concurrently?

File: go.sum
Comment (nsrip-dd) [inline go.sum:803]: These changes don't seem to be needed. I'm guessing they're left over from local work? If I remove these lines and run `go mod tidy`, they don't come back, so it should be fine to delete them.

**Review by nsrip-dd — DISMISSED** [review]

File: internal/globalconfig/globalconfig.go
Comment (katiehockman) [inline internal/globalconfig/globalconfig.go:77]: This is only used in tests, correct? Is it necessary to have it part of the API in this way, or can this just be directly in the tests where it's needed?

File: internal/normalizer/normalizer.go
Comment (katiehockman) [inline internal/normalizer/normalizer.go:22]:
```suggestion
// e.g, "first:second:third" gets split into `header = "first:second"` and `tag="third"`
```

File: contrib/emicklei/go-restful/option.go
Comment (katiehockman) [inline contrib/emicklei/go-restful/option.go:70]: From the RFC:

> In each integration docs MUST be made clear that for special headers we don't sub-select
> values. E.g. Cookie header can appear multiple times for different cookies that have to be set.
> The cookie name appears in the value.
> We DO NOT provide a way to only select a specific cookie. If a user activates the header
> Cookie then ALL cookies will be transmitted.

So let's be sure to document that in every integration. You can check out the implementation from other languages to see the wording they used if you'd like.

Comment (katiehockman) [reply contrib/emicklei/go-restful/option.go:70]: Let's also note that setting the integration-specific header tags option in this way will overwrite any existing global header tags configurations that have previously been set.

**Review by katiehockman — APPROVED** [review]
Just a few other comments but approved once those are addressed! Thanks for your patience!!

File: internal/globalconfig/globalconfig.go
Comment (mtoffl01) [reply internal/globalconfig/globalconfig.go:77]: Just to make sure we're on the same page, having a public getter for the config can be a problem due to this: https://github.com/DataDog/dd-trace-go/pull/1764#discussion_r1172670466, right?

File: internal/globalconfig/globalconfig.go
Comment (mtoffl01) [reply internal/globalconfig/globalconfig.go:28]: @katiehockman , when you get the chance

File: contrib/emicklei/go-restful/option.go
Comment (mtoffl01) [reply contrib/emicklei/go-restful/option.go:76]: Each integration has a function that runs to configure the default config object (on some integrations, its called `newConfig`, `defaults` for others).
Regardless of the name, they all set cfg.headersAsTag to globalconfig.GetAllHeaderTags(), which "returns a copy of the header tags held in the global config or the empty map"

File: internal/globalconfig/globalconfig.go
Comment (mtoffl01) [inline internal/globalconfig/globalconfig.go:94]: @katiehockman , in the case that a user configured `DD_TRACE_HEADER_TAGS` and `WithHeaderTags()`, the contrib option should take precedence. So it needs to wipe out the config that from the env var, hence the need for this function.

File: internal/globalconfig/globalconfig_test.go
Comment (katiehockman) [inline internal/globalconfig/globalconfig_test.go:14]:
```suggestion
func TestHeaderTag(t *testing.T) {
```

**Review by katiehockman — COMMENTED** [review]
Looks good! Only thing I'm concerned by is that this changes service names. Maybe there was a bad rebase or something

File: contrib/net/http/option.go
Comment (katiehockman) [inline contrib/net/http/option.go]: I don't think you meant to drop this :)

**Review by katiehockman — APPROVED** [review]
LGTM aside from that last comment

**Review by katiehockman — APPROVED** [review]

File: internal/normalizer/normalizer.go
Comment (Julio-Guerra) [reply internal/normalizer/normalizer.go:26]: my bad, just saw this wasn't use with the request headers but rather with the user config. resolving this

**Review by Julio-Guerra — APPROVED** [review]

File: internal/normalizer/normalizer.go
Comment (mtoffl01) [reply internal/normalizer/normalizer.go:26]: Yes, header names cannot contain `:` in HTTP1: https://www.rfc-editor.org/rfc/rfc7230#section-3.2
HTTP2 includes “pseudo-headers” whose header names are prefixed with a colon, see: https://httpwg.org/specs/rfc7540.html#HttpHeaders. This implies there is at least some level of support for colons in header names in HTTP2, but I was unable to determine whether HTTP2 allows custom header names with colons. 

This feature was written with HTTP1 in mind.

**Pull request merged**
